### ISSUE TITLE
STOP_LOSS and TAKE_PROFIT 

### ISSUE THREAD

**Issue by namato18** [datawookie/binance#17]
I see that "TAKE_PROFIT" and "STOP_LOSS" are options for the 'order_type' argument of the 'spot_new_order()' function. However, I can't seem to successfully place one of these orders. I've tried adding the 'stopPrice' argument, but that doesn't seem to help.

I tried running a test using this code:

```
spot_new_order(
  order_type = 'STOP_LOSS',
  symbol = 'BTCUSDT',
  side = 'SELL',
  stopPrice = 20000,
  test = TRUE
)
```

Any help or examples would be greatly appreciated as I really enjoy using this package so far!

**Comment by datawookie** [comment]
Hi Nick,

You mentioned that you attempted several modifications which did not work. It would be helpful if you provided some information on those. This will help me avoid re-inventing the wheel and mean that I am able to resolve your problem more quickly. Please can you add these here.

Can you also please clarify precisely what you are trying to do? For example, referring to your code fragment above, do you already have an open order?

Finally, have you checked to see whether the `STOP_LOSS` order type is supported on the exchange and currency pair you are considering?

```
> market_exchange_info()$symbols %>% filter(symbol == "BTCUSDT") %>% pull(order_types)
[[1]]
[1] "LIMIT"             "LIMIT_MAKER"       "MARKET"            "STOP_LOSS_LIMIT"   "TAKE_PROFIT_LIMIT"
[[2]]
[1] "LIMIT"             "LIMIT_MAKER"       "MARKET"            "STOP_LOSS_LIMIT"   "TAKE_PROFIT_LIMIT"
[[3]]
[1] "LIMIT"             "LIMIT_MAKER"       "MARKET"            "STOP_LOSS_LIMIT"   "TAKE_PROFIT_LIMIT"
[[4]]
[1] "LIMIT"             "LIMIT_MAKER"       "MARKET"            "STOP_LOSS_LIMIT"   "TAKE_PROFIT_LIMIT"
```

Regards, Andrew.

**Comment by namato18** [comment]
Hi Andrew,

Sorry about that, I wasn't using the allowed 'STOP_LOSS_LIMIT' for BTCUSDT like I should have been instead of 'STOP_LOSS'. As a more detailed example, let's say I buy 1 BTC using USDT and a MARKET order:

spot_new_order(
  order_type = "MARKET",
  symbol = "BTCUSDT",
  side = "BUY",
  quantity = 1,
  test = TRUE
)

The trade goes through successfully and I now own 1 BTC. Now I wish to set a stop loss. To do this, I've attempted to use the following code: 

spot_new_order(
  order_type = 'STOP_LOSS_LIMIT',
  symbol = 'BTCUSDT',
  side = 'SELL',
  quantity = 1,
  price = 25000,
  time_in_force = 'GTC',
  test = TRUE,
)

The way I would read this is that when BTCUSDT reaches a price of 25000 I will sell 1 BTC and get USDT in return for that trade. Perhaps I'm misunderstanding the concept. This code returns the following error:

Error: Binance API request failed [400] At least one of 'stopPrice' and 'trailingDelta' was not sent.

Perhaps I am misunderstanding since I thought that the 'price' argument in the code above would specify the stop price. It could be that the 'price' argument is specifying where to place a limit order and the 'stopPrice' is the stopping price of that limit order? I tried adding the stop price as follows:

spot_new_order(
  order_type = 'STOP_LOSS_LIMIT',
  symbol = 'BTCUSDT',
  side = 'SELL',
  quantity = 1,
  price = 25000,
  stopPrice = 24000,
  time_in_force = 'GTC',
  test = TRUE,
)

which returns the following error:

Error in spot_new_order(order_type = "STOP_LOSS_LIMIT", symbol = "BTCUSDT",  :  unused argument (stopPrice = 24000)

In any case, I'm unsure how to add either 'stopPrice' or 'trailingDelta' as an argument to the function spot_new_order().

**Comment by datawookie** [comment]
Hi @namato18,

I have added support for the stop loss and trailing stop parameters. Can you help me test please? You'll need to install from the `stop-take-profit` branch.

Then try something like this:

```
order <- spot_new_order(
  order_type = "STOP_LOSS_LIMIT",
  symbol = "ETHBUSD",
  side = "SELL",
  quantity = 1,
  price = 2200,
  time_in_force = "GTC",
  stop_price = 1700
)
```

Let me know how that goes.

Thanks, Andrew.

**Comment by namato18** [comment]
Hi @datawookie,

Using the updated 'stop-take-profit' branch, the function spot_new_order() returned TRUE when I did a test run. I also tested it using the following code and it worked fine:

```
order <- spot_new_order(
  order_type = "STOP_LOSS_LIMIT",
  symbol = "BTCUSDT",
  side = "SELL",
  quantity = 1,
  price = 22000,
  time_in_force = "GTC",
  stop_price = 17000,
  test = TRUE
)
```

So it appears to be working. Am I correct in assuming that the 'price' is where the limit order will be placed and the 'stop_price' is where the limit order will fill?

Thanks for such quick updates!

Nick

**Comment by namato18** [comment]
I did a live test using DOGECOIN. I bought DOGECOIN using the following code:
```
spot_new_order(
  order_type = "MARKET",
  symbol = "DOGEUSDT",
  side = "BUY",
  quantity = 100,
)
```
So now I own ~100 DOGECOIN. Next I placed a stop loss using an order type of 'STOP_LOSS_LIMIT'. This worked as well! So I was misunderstanding how it would function a little bit, but that's probably my fault. The 'stop_price' is the limit orders trigger price (if DOGEUSDT reaches this price it will try to fill your order). 'price' is the value at which your order will try to fill at. This was the code I tested it with:
```
order <- spot_new_order(
  order_type = "STOP_LOSS_LIMIT",
  symbol = "DOGEUSDT",
  side = "SELL",
  quantity = 99,
  price = 0.055,
  time_in_force = "GTC",
  stop_price = 0.050,
)
```
Meaning my limit order would trigger at 0.050 and it would try to sell my 99 DOGECOIN at 0.055 USDT each. Would this order fill?? Possibly, but it's probably best to have your price close to your stop_price to avoid unfilled orders.

I tested the 'TAKE_PROFIT_LIMIT' order type as well, it works in a similar fashion, but just reverse. Hopefully this small example can help out anyone else in the future!

I really appreciate the quick updates @datawookie!
